src/object.jl

```diff
@@ -186,21 +192,45 @@ mutable struct PrefechedDownloadStream{T <: Object} <: IO
         len = parse(Int, HTTP.header(resp, "Content-Length", "0"))
         et = etag(HTTP.header(resp, "ETag", ""))
         object = Object(store, credentials, String(key), Int(len), String(et))
-        return PrefechedDownloadStream(object, prefetch_size; prefetch_multipart_size)
+        return PrefetchedDownloadStream(object, prefetch_size; prefetch_multipart_size)
     end
 end
-Base.eof(io::PrefechedDownloadStream) = io.pos >= io.len
-Base.bytesavailable(io::PrefechedDownloadStream) = io.len - io.pos + 1
-function Base.close(io::PrefechedDownloadStream)
+Base.eof(io::PrefetchedDownloadStream) = io.pos >= io.len
+bytesremaining(io::PrefetchedDownloadStream) = io.len - io.pos + 1
+function Base.bytesavailable(io::PrefetchedDownloadStream)
+    return isnothing(io.buf) ? 0 : bytesavailable(io.buf::PrefetchBuffer)
+end
+function Base.close(io::PrefetchedDownloadStream)
     close(io.prefetch_queue)
     close(io.download_queue)
-    Base.notify_error(io.cond.cond_wait, Base.closed_exception())
+    Base.@lock io.cond.cond_wait begin
+        Base.notify_error(io.cond.cond_wait, Base.closed_exception())
+    end
+    if !isnothing(io.buf)
+        resize!(io.buf.data, 0)
+        io.buf = nothing
+    end
     return nothing
 end
-Base.isopen(io::PrefechedDownloadStream) = isopen(io.prefetch_queue)
-Base.filesize(io::PrefechedDownloadStream) = io.len
+Base.isopen(io::PrefetchedDownloadStream) = !(isnothing(io.buf) && !isopen(io.prefetch_queue))
+Base.iswritable(io::PrefetchedDownloadStream) = false
+Base.filesize(io::PrefetchedDownloadStream) = io.len
+function Base.peek(io::PrefetchedDownloadStream, ::Type{T}) where {T<:Integer}
+    eof(io) && throw(EOFError())
+    buf = getbuffer(io)
+    # TODO: allow cross-buffer peeking
+    bytesavailable(buf) < sizeof(T) && error(string(
+        "Cannot peak ahead $(sizeof(T)) bytes for T=$T, there are only",
+        "$(bytesavailable(buf)) bytes in the current buffer."
+    ))
+    GC.@preserve buf begin
+        ptr::Ptr{T} = pointer(buf.data, buf.pos)
+        x = unsafe_load(ptr)
```

Comment:
Why do we need this instead of just `buf.data[buf.pos]`? do we really have binary data in buf that we can unsafe_load?

Reply:
I copy-pasted the code from [IOBuffer](https://github.com/JuliaLang/julia/blob/master/base/iobuffer.jl#L175-L187), we need to do this to cast the underlying data to a wider type like UInt16. There are test-cases for this in the bottom of `runtests.jl`, I think this should be as safe as peeking an `IOBuffer`, unless I'm missing something.

Reply:
Ah, ok; yeah, that's fine. I was more curious in our _own_ use-case if you had a case where we were wanting to peek at something that wasn't a `UInt8`.